pstoremem/metadata.go

```diff
@@ -7,37 +7,37 @@ import (
 	pstore "github.com/libp2p/go-libp2p-peerstore"
 )
 
+type metakey struct {
+	id  peer.ID
+	key string
+}
+
 type memoryPeerMetadata struct {
 	// store other data, like versions
 	//ds ds.ThreadSafeDatastore
-	ds     map[string]interface{}
+	ds     map[metakey]interface{}
 	dslock sync.Mutex
 }
 
 var _ pstore.PeerMetadata = (*memoryPeerMetadata)(nil)
 
 func NewPeerMetadata() pstore.PeerMetadata {
 	return &memoryPeerMetadata{
-		ds: make(map[string]interface{}),
+		ds: make(map[metakey]interface{}),
 	}
 }
 
 func (ps *memoryPeerMetadata) Put(p peer.ID, key string, val interface{}) error {
-	//dsk := ds.NewKey(string(p) + "/" + key)
-	//return ps.ds.Put(dsk, val)
 	ps.dslock.Lock()
 	defer ps.dslock.Unlock()
-	ps.ds[string(p)+"/"+key] = val
+	ps.ds[metakey{p, key}] = val
```

Comment:
how does this work with the custom struct? Wouldn't it create a new mapping? There doesn't seem to be any methods on the struct to support map operations.

Reply:
I mean, this won't deduplicate if that was the intention.

Reply:
> how does this work with the custom struct? Wouldn't it create a new mapping? There doesn't seem to be any methods on the struct to support map operations.

Maps work with struct keys as long as the map works with all the fields of the struct: https://blog.golang.org/go-maps-in-action#TOC_5.

> I mean, this won't deduplicate if that was the intention.

It saves us from storing an additional copy of the peer ID concatenated with the metadata key. Instead, we just keep a pointer to the peer ID and the metadata key.

Reply:
This should dedup, as we're storing the struct value, not a pointer. From the go spec:

> Struct values are comparable if all their fields are comparable. Two struct values are equal if their corresponding non-blank fields are equal.

Also see: https://blog.golang.org/go-maps-in-action#TOC_5.

Reply:
Sure, I can see the benefit. I just didn't know that maps work this way with structs.

Reply:
Yeah, go maps are very magical...

Reply:
You had already got to it by the time I released the review :P